airbyte-webapp/src/components/StateBlock/StateBlock.tsx

```diff
@@ -0,0 +1,168 @@
+import Editor from "react-simple-code-editor";
+import styled from "styled-components";
+
+import { H5, LoadingButton } from "components";
+import ContentCard from "components/ContentCard";
+
+import { ConnectionState, ConnectionStateObject } from "core/request/AirbyteClient";
+import { useConfirmationModalService } from "hooks/services/ConfirmationModal";
+import { useGetConnectionState, useSetConnectionState } from "hooks/services/useConnectionHook";
+
+type IProps = {
+  connectionId: string;
+};
+
+const StateBlockComponent = styled(ContentCard)`
+  margin-top: 12px;
+  padding: 19px 20px 20px;
+  // display: flex;
+  align-items: top;
+  justify-content: space-between;
+`;
+
+const Text = styled.div`
+  margin-left: 20px;
+  font-size: 11px;
+  line-height: 13px;
+  color: ${({ theme }) => theme.greyColor40};
+  white-space: pre-line;
+`;
+
+const ErrorText = styled.div`
+  margin-left: 20px;
+  font-size: 11px;
+  line-height: 13px;
+  color: ${({ theme }) => theme.redColor};
+  white-space: pre-line;
+`;
+
+const codeStyle = {
+  fontFamily: '"Fira code", "Fira Mono", monospace',
+  fontSize: 12,
+  marginTop: "5px",
+  marginLeft: "10px",
+  marginBottom: "5px",
+};
+
+type Validation = {
+  valid: boolean;
+  message?: string;
+};
+
+const StateBlock: React.FC<IProps> = ({ connectionId }) => {
+  const [stateString, setStateString] = useState<string>(`// ...`);
+  const [loading, setLoading] = useState(false);
+  const [validation, setValidation] = useState<Validation>({ valid: true });
+  const { openConfirmationModal, closeConfirmationModal } = useConfirmationModalService();
+  const { mutateAsync: getState } = useGetConnectionState();
+  const { mutateAsync: setState } = useSetConnectionState();
+
+  const loadState = async () => {
+    setLoading(true);
+    const state = await getState(connectionId);
+    if (state?.state) {
+      setStateString(formatState(state.state));
+      setLoading(false);
+    }
+  };
+
+  const saveState = async () => {
+    setLoading(true);
+    const stateObject = JSON.parse(stateString) as ConnectionState;
+    const newState = await setState({ connectionId, state: stateObject });
+    if (newState?.state) {
+      setStateString(formatState(newState.state));
+      setLoading(false);
+    }
+  };
+
+  const onSaveButtonClick = useCallback(() => {
+    const validationResponse = validateState(stateString);
+    setValidation(validationResponse);
+    if (!validationResponse.valid) return;
+
+    openConfirmationModal({
+      text: `tables.State.confirmModalText`,
+      title: `tables.State.confirmModalTitle`,
+      submitButtonText: "form.save",
+      onSubmit: async () => {
+        await saveState();
+        closeConfirmationModal();
+      },
+      submitButtonDataId: "state",
+    });
+  }, [closeConfirmationModal, saveState, openConfirmationModal]);
+
+  const loadStateMemoized = useCallback(() => {
+    loadState();
+    // eslint-disable-next-line
+  }, []);
```

Comment:
Any tricks to memoize a method that takes no args?

Reply:
There is no need to memoize this method at all. This method is supposed to be only called once on initialization in the `useEffect` below. It's not passed in to any component that would suddenly rerender unnecessarily by havnig a changed reference here. So this is actually atm only having a performance impact (for the comparison of the useCallback dependencies) on every render, and no benefit.

An additional side-node. the `useMemo` (and therefore `useCallback`) specificaitions are meant to do only performance optimizations, i.e. they don't give a "guarantee" that they won't change despite the dependency array haven't changed. Which - if this would ever change - would mean that we actually call it again in the `useEffect` below and thus actually have a breaking behavior.